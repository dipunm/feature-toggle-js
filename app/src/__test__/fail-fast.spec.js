const FeatureToggles = require('../server');

describe('param validation when calling FeatureToggles.create()', () => {
  const someFeatures = [
    { name: 'toggle1', test: () => true },
    { name: 'toggle2', test: () => true },
    { name: 'toggle3', test: () => true },
  ];
  function change(arr, ind, mutate) {
    return arr.map((e, i) => {
      if (i === ind) {
        const obj = Object.assign({}, e);
        mutate(obj);
        return obj;
      }
      return e;
    });
  }

  const badArgumentSpecs = (() => {
    const specs = [];
    specs.push(['given no parameters', () => FeatureToggles.create()]);
    specs.push(['given an empty array', () => FeatureToggles.create([])]);

    specs.push([
      'given a toggle object with no name',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          delete toggle.name;
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with an empty name',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.name = '';
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a non-string name',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.name = 0;
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a duplicate name',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.name = 'toggle1';
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with no test',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          delete toggle.test;
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a non function test',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.test = true;
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a non function test',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.test = true;
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a non array dependencies',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.dependencies = 'env';
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a non function health',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.health = 'unhealthy';
        });
        FeatureToggles.create(features);
      },
    ]);

    specs.push([
      'given a toggle object with a non function resetOn',
      () => {
        const features = change(someFeatures, 1, (toggle) => {
          toggle.resetOn = '2h';
        });
        FeatureToggles.create(features);
      },
    ]);
    return specs;
  })();

  describe.each(badArgumentSpecs)('|| %s ||', (desc, action) => {
    test('should throw with a meaningful message', () => {
      expect(action).toThrowErrorMatchingSnapshot();
    });
  });
});
