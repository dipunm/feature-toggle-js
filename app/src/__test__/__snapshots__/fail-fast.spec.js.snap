// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a duplicate name || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" position 1 contains a duplicate value"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a non array dependencies || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"dependencies\\" fails because [\\"dependencies\\" must be an array]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a non function health || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"health\\" fails because [\\"health\\" must be a Function]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a non function resetOn || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"resetOn\\" fails because [\\"resetOn\\" must be a Function]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a non function test || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"test\\" fails because [\\"test\\" must be a Function]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a non function test || should throw with a meaningful message 2`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"test\\" fails because [\\"test\\" must be a Function]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with a non-string name || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"name\\" fails because [\\"name\\" must be a string]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with an empty name || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"name\\" fails because [\\"name\\" is not allowed to be empty]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with no name || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"name\\" fails because [\\"name\\" is required]]"`;

exports[`param validation when calling FeatureToggles.create() || given a toggle object with no test || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" at position 1 fails because [child \\"test\\" fails because [\\"test\\" is required]]"`;

exports[`param validation when calling FeatureToggles.create() || given an empty array || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" must contain at least 1 items"`;

exports[`param validation when calling FeatureToggles.create() || given no parameters || should throw with a meaningful message 1`] = `"Argument exception: \\"features\\" is required"`;
